aten/src/ATen/native/Histogram.cpp

```diff
@@ -0,0 +1,211 @@
+    TORCH_CHECK(input.dtype() == hist.dtype(), "torch.histogram: input tensor and hist tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and hist ", hist.dtype());
+
+    TORCH_CHECK(input.dtype() == bin_edges.dtype(), "torch.histogram: input tensor and bin_edges tensor should",
+            " have the same dtype, but got input ", input.dtype(), " and bin_edges ", bin_edges.dtype());
+
+    TORCH_CHECK(bin_ct > 0,
+            "torch.histogram(): bins must be > 0, but got ", bin_ct);
+
+    at::native::resize_output(hist, bin_ct);
+
+    at::native::resize_output(bin_edges, bin_ct + 1);
+
+    TORCH_CHECK(hist.is_contiguous(), "torch.histogram: hist tensor must be contiguous");
+}
+
+/* Determines the outermost bin edges.
+ */
+std::pair<double, double> select_outer_bin_edges(const Tensor& input,
+        const c10::optional<Scalar>& min, const c10::optional<Scalar>& max) {
+    if (min.has_value() && max.has_value()) {
+        return std::make_pair(min.value().to<double>(), max.value().to<double>());
+    }
+
+    // Default range for empty input matching numpy.histogram's default
+    if (input.numel() == 0) {
+        return std::make_pair(0., 1.);
+    }
+
+    auto extrema = _aminmax(input);
+
+    double leftmost_edge  = min.has_value() ? min.value().to<double>() : std::get<0>(extrema).item<double>();
+    double rightmost_edge = max.has_value() ? max.value().to<double>() : std::get<1>(extrema).item<double>();
+
+    TORCH_CHECK(leftmost_edge <= rightmost_edge, "torch.histogram: min should not exceed max, but got",
+            "min ", leftmost_edge, " max ", rightmost_edge);
+
+    // Expand empty range to match numpy behavior and avoid division by 0 in normalization
+    if (leftmost_edge == rightmost_edge) {
+        leftmost_edge -= 0.5;
+        rightmost_edge += 0.5;
+    }
+
+    return std::make_pair(leftmost_edge, rightmost_edge);
+}
+
+/* histc's version of the logic for outermost bin edges.
+ */
+std::pair<double, double> histc_select_outer_bin_edges(const Tensor& input,
+		const Scalar& min, const Scalar& max) {
+	double leftmost_edge = min.to<double>();
+	double rightmost_edge = max.to<double>();
+
+    if (leftmost_edge == rightmost_edge) {
+        auto extrema = _aminmax(input);
+        leftmost_edge = std::get<0>(extrema).item<double>();
+        rightmost_edge = std::get<1>(extrema).item<double>();
+    }
+
+    if (leftmost_edge == rightmost_edge) {
+        leftmost_edge -= 1;
+        rightmost_edge += 1;
+    }
+
+    TORCH_CHECK(!(std::isinf(leftmost_edge) || std::isinf(rightmost_edge) ||
+            std::isnan(leftmost_edge) || std::isnan(rightmost_edge)),
+            "range of [", leftmost_edge, ", ", rightmost_edge, "] is not finite");
+
+    TORCH_CHECK(leftmost_edge < rightmost_edge, "torch.histc: max must be larger than min");
+
+    return std::make_pair(leftmost_edge, rightmost_edge);
+}
+
+} // namespace
+
+std::tuple<Tensor&, Tensor&>
```

Comment:
Add comments explaining how each of these groupings is distinct. In this case, for example, bins is a tensor